ceci/config.py

```diff
@@ -0,0 +1,233 @@
+        default : `dtype` or `None`
+            The default value
+        format : `str`
+            A formatting string for printout and representation
+        help : `str`
+            A help or docstring
+        """
+        kwcopy = kwargs.copy()
+        self._help = kwcopy.pop('help', 'A Parameter')
+        self._format = kwcopy.pop('format', '%s')
+        self._dtype = kwcopy.pop('dtype', None)
+        self._default = kwcopy.pop('default', None)
+        if kwcopy:  # pragma: no cover
+            raise ValueError(f"Unknown arguments to StageParameter {str(list(kwcopy.keys()))}")
+        self._value = cast_value(self._dtype, self._default)
+
+    @property
+    def value(self):
+        """ Return the value """
+        return self._value
+
+    @property
+    def dtype(self):
+        """ Return the data type """
+        return self._dtype
+
+    @property
+    def default(self):
+        """ Return the default value """
+        return self._default
+
+    def set(self, value):
+        """ Set the value, raising a TypeError if the value is the wrong type """
+        self._value = cast_value(self._dtype, value)
+        return self._value
+
+    def set_to_default(self):
+        """ Set the value to the default """
+        self._value = cast_value(self._dtype, self._default)
+        return self._value
+
+
+class StageConfig(dict):
+    """ A small class to manage a dictionary of configuration parameters with basic type checking
+    """
+
+    def __init__(self, **kwargs):
+        """ Build from keywords
+
+        Note
+        ----
+        The keywords are used as keys for the configuration parameters
+
+        The values are used to define the allowed data type and default values
+
+        For each key-value pair:
+        If the value is a type then it will define the data type and the default will be `None`
+        If the value is a value then it will set the default value define the data type as type(value)
+        """
+        dict.__init__(self)
+        for key, val in kwargs.items():
+            if val is None:
+                dtype = None
+                default = None
+            elif isinstance(val, type):
+                dtype = val
+                default = None
+            else:
+                dtype = type(val)
+                default = val
+            param = StageParameter(dtype=dtype, default=default)
+            self[key] = param
+
+    def __str__(self):
+        """ Override __str__ casting to deal with `StageParameter` object in the map """
+        s = "{"
+        for key, attr in self.items():
+            if isinstance(attr, StageParameter):
+                val = attr.value
+            else:
+                val = attr
+            s += f"{key}:{val},"
+        s += "}"
+        return s
+
+    def __repr__(self):
+        """ A custom representation """
+        s = "StageConfig"
+        s += self.__str__()
+        return s
+
+    def __getitem__(self, key):
```

Comment:
Will this also work when iterating with `items()` and call `attr.value` there too?

Reply:
Good question, I'll look into this.

Reply:
No, it doesn't:

```
>>> from ceci.config import StageConfig
>>> config_options = {'chunk_rows': 5000, 'something':float, 'free':None}
>>> config = StageConfig(**config_options)
>>> for k, v in config.items():
...   print(k,v)
... 
chunk_rows <ceci.config.StageParameter object at 0x7f9d870cfd90>
something <ceci.config.StageParameter object at 0x7f9d85e60eb0>
free <ceci.config.StageParameter object at 0x7f9d870a6610>

```

We could override `items() and values()` to do this, or we could leave it as a way to talk to the paramters if needed, thoughts?

Reply:
It would be nice to be able to do that, to give stages something that's as dictionary-like as possible.

Reply:
Ok, I'm not seeing a way to do that that is clearly exactly what we want.  `items()` returns `dict_items` and `values()` returns `dict_values`, which are builtin classes that I can't seem to use.  Probably the best is if return a list of tuple(key, value) for items() and a list for values().    

There are other patterns that allow you to control the type of object that you are allowed to assign to data members, but they are somewhat more complicated and hard to understand, overriding `__get__` and `__set__` and hiding attributes in the parent class.   My vote would be just to return lists.   I'll go ahead and do that.

Reply:
I did this:

```
    def items(self):
        """ Override items() to get the parameters values instead of the objects """
        return [(key, cast_to_streamable(value)) for key, value in dict.items(self)]

    def values(self):
        """ Override items() to get the parameters values instead of the objects """
        return [cast_to_streamable(value) for value in dict.values(self)]
```

Reply:
I think that might not be quite as efficient for doing things like unions & other complicated operations, but should work fine.